src/data_race.rs

```diff
@@ -542,6 +542,36 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
         Ok(old)
     }
 
+    /// Perform an conditional atomic exchange with a memory place and a new
+    /// scalar value, the old value is returned.
+    fn atomic_conditional_exchange_scalar(
+        &mut self,
+        place: MPlaceTy<'tcx, Tag>,
+        rhs: ImmTy<'tcx, Tag>,
+        min: bool,
+        atomic: AtomicRwOp,
+    ) -> InterpResult<'tcx, ImmTy<'tcx, Tag>> {
+        let this = self.eval_context_mut();
+
+        let old = this.allow_data_races_mut(|this| this.read_immediate(place.into()))?;
+        this.validate_atomic_rmw(place, atomic)?;
+
+        // `binary_op` will bail if either of them is not a scalar.
+        let cond = if min {
+            this.overflowing_binary_op(mir::BinOp::Lt, old, rhs)?.0
+        } else {
+            this.overflowing_binary_op(mir::BinOp::Gt, old, rhs)?.0
+        };
+
+        if cond.to_bool()? {
+            this.allow_data_races_mut(|this| this.write_immediate(*rhs, place.into()))?;
+            this.validate_atomic_rmw(place, atomic)?;
```

Comment:
The second validate_atomic_rmw is unnecessary, i think atomic max & min provides the same semantics regarding ordering even if the value is not changed? (either way the first validate means this would just increment the counters), I would also move the first validate_atomic_rmw to the end of the function for consistency with the other functions as well.

Reply:
I assume if the value is not changed, then like CmpXchg this is only considered a read, not a write?

Reply:
For the two architecures i checked, LLVM compiles AtomicUsize::fetch_max(0, Ordering::AcqRel) to an AtomicUsize::fetch_or(0, Ordering::AcqRel) or similar style atomic ops for the two architectures i checked which supports my interpretation. It would also be unusual for fetch_max to have different semantics to fetch_add and not continue release sequences even if no change occurs.

Reply:
For example both of these functions in LLVM:
```
pub fn null_max_release_u8(x: &AtomicU8) -> u8 {
    x.fetch_max(0, Ordering::Release)
}

pub fn null_max_release_usize(x: &AtomicUsize) -> usize {
    x.fetch_max(0, Ordering::Release)
}
```
compile to (for Ordering::Release, Ordering::AcqRel, Ordering::SeqCst):
```
        mfence
        mov     rax, qword ptr [rdi]
        ret
```

Reply:
So adding one atomic semantic check at the end is what we need here, I was confused a little on the exact semantics coming from LLVM.

Reply:
I was not concerned about the interaction with release sequences, but about the notion of data races. If a non-atomic access happens concurrently with a non-mutating RMW, is this a data race or not? I thought the answer was "no".

For release sequences (or release semantics in general)... an AcqRel RMW uses Acq semantics for the "unsuccessful" memory ordering, right? So there's no release happening when the compare-exchange fails.

OTOH... there is no such thing as a "failing" fetch_or / fetch_add / fetch_max. So my analogy makes no sense. I retract my objection -- all the "never-failing" `fetch_*` operations should clearly behave the same, and they should be "release" even if they do not change the value stored at that location. But I expect this also means they are always considered to race with concurrency non-atomic reads, even if they do not change the value stored at that location.

Reply:
FWIW, the [stable docs](https://doc.rust-lang.org/1.48.0/std/sync/atomic/struct.AtomicI32.html#method.fetch_max) for `fetch_max` say:

> Finds the maximum of the current value and the argument val, and sets the new value to the result.

My reading of that sentence is that the operation is "read then compute then store" unconditionally like the other `fetch_*`.